<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Database-encoding-with-basic-search" data-toc-modified-id="Database-encoding-with-basic-search-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Database encoding with basic search</a></span><ul class="toc-item"><li><span><a href="#2,2-register-system-(hand-encoded),-with-data-array-[3,2,0,1]" data-toc-modified-id="2,2-register-system-(hand-encoded),-with-data-array-[3,2,0,1]-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>2,2 register system (hand encoded), with data array [3,2,0,1]</a></span></li><li><span><a href="#3,4-registers-building-(hand-encoded),-with-data-array-[3,4,9,13]" data-toc-modified-id="3,4-registers-building-(hand-encoded),-with-data-array-[3,4,9,13]-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>3,4 registers building (hand encoded), with data array [3,4,9,13]</a></span></li><li><span><a href="#Automatic-bitstring-encoding" data-toc-modified-id="Automatic-bitstring-encoding-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>Automatic bitstring encoding</a></span></li></ul></li><li><span><a href="#Bit-String-Comparator" data-toc-modified-id="Bit-String-Comparator-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Bit String Comparator</a></span></li><li><span><a href="#Bit-String-Comparator-inside-Grover" data-toc-modified-id="Bit-String-Comparator-inside-Grover-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Bit String Comparator inside Grover</a></span><ul class="toc-item"><li><span><a href="#Creating-functions-to-make-the-circuit-scalable-to-different-sized-bitstrings" data-toc-modified-id="Creating-functions-to-make-the-circuit-scalable-to-different-sized-bitstrings-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Creating functions to make the circuit scalable to different sized bitstrings</a></span><ul class="toc-item"><li><span><a href="#Here-we-build-the-circuit-that-will-adapt-depending-on-the-size-of-the-bitstrings-to-compare" data-toc-modified-id="Here-we-build-the-circuit-that-will-adapt-depending-on-the-size-of-the-bitstrings-to-compare-3.1.1"><span class="toc-item-num">3.1.1&nbsp;&nbsp;</span>Here we build the circuit that will adapt depending on the size of the bitstrings to compare</a></span></li></ul></li><li><span><a href="#Making-the-Grover-Iteration-into-a-function-to-be-able-to-apply-it-multiple-times-to-lower-the-prob.-of-wrong-answers-(same-comparison-as-two-above)" data-toc-modified-id="Making-the-Grover-Iteration-into-a-function-to-be-able-to-apply-it-multiple-times-to-lower-the-prob.-of-wrong-answers-(same-comparison-as-two-above)-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Making the Grover Iteration into a function to be able to apply it multiple times to lower the prob. of wrong answers (same comparison as two above)</a></span></li></ul></li><li><span><a href="#Assembling-both-comparator-and-database-into-one-(final-code)" data-toc-modified-id="Assembling-both-comparator-and-database-into-one-(final-code)-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Assembling both comparator and database into one (final code)</a></span></li><li><span><a href="#Experimenting-with-Backends" data-toc-modified-id="Experimenting-with-Backends-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Experimenting with Backends</a></span><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#Links-and-ressources-that-I-found" data-toc-modified-id="Links-and-ressources-that-I-found-5.0.1"><span class="toc-item-num">5.0.1&nbsp;&nbsp;</span>Links and ressources that I found</a></span></li></ul></li><li><span><a href="#Testing-Area" data-toc-modified-id="Testing-Area-5.1"><span class="toc-item-num">5.1&nbsp;&nbsp;</span>Testing Area</a></span></li><li><span><a href="#Characterising--the-circuit" data-toc-modified-id="Characterising--the-circuit-5.2"><span class="toc-item-num">5.2&nbsp;&nbsp;</span>Characterising  the circuit</a></span><ul class="toc-item"><li><span><a href="#Creating-the-dataset" data-toc-modified-id="Creating-the-dataset-5.2.1"><span class="toc-item-num">5.2.1&nbsp;&nbsp;</span>Creating the dataset</a></span></li><li><span><a href="#Plot-Using-go.Surface()" data-toc-modified-id="Plot-Using-go.Surface()-5.2.2"><span class="toc-item-num">5.2.2&nbsp;&nbsp;</span>Plot Using go.Surface()</a></span></li><li><span><a href="#Plot-Using-go.Heatmap()" data-toc-modified-id="Plot-Using-go.Heatmap()-5.2.3"><span class="toc-item-num">5.2.3&nbsp;&nbsp;</span>Plot Using go.Heatmap()</a></span><ul class="toc-item"><li><span><a href="#Discrete" data-toc-modified-id="Discrete-5.2.3.1"><span class="toc-item-num">5.2.3.1&nbsp;&nbsp;</span>Discrete</a></span></li><li><span><a href="#Interpolated-(smoothed)" data-toc-modified-id="Interpolated-(smoothed)-5.2.3.2"><span class="toc-item-num">5.2.3.2&nbsp;&nbsp;</span>Interpolated (smoothed)</a></span></li></ul></li><li><span><a href="#Plot-Using-px.imshow()-for-heatmap" data-toc-modified-id="Plot-Using-px.imshow()-for-heatmap-5.2.4"><span class="toc-item-num">5.2.4&nbsp;&nbsp;</span>Plot Using px.imshow() for heatmap</a></span></li></ul></li></ul></li></ul></div>

In [174]:
from qiskit import QuantumCircuit, assemble, Aer, QuantumRegister, ClassicalRegister, AncillaRegister, transpile
from qiskit.visualization import plot_bloch_multivector, plot_histogram, array_to_latex
import os
import math
from math import log, ceil, floor
import plotly.express as px
import plotly.graph_objects as go
from plotly.subplots import make_subplots
import numpy as np
from random import randrange

# Database encoding with basic search
## 2,2 register system (hand encoded), with data array [3,2,0,1]

The value to search is also hand encoded.

In [91]:
# Making the circuit in the paper

n = 2 #number of index qbits

qrd = QuantumRegister(n, 'd') #Creates the data registry
qrt = QuantumRegister(n, 't') #Creates the template registry
anc = QuantumRegister(1, 'ancilla') #Creates the ancillary registry for the oracle
cr = ClassicalRegister(n, 'c') #Creates the classical bit measurment output

qc = QuantumCircuit(qrd, qrt, anc, cr) #Makes the circuit with these qbits as input

####### State initialization #########

qc.h(range(n))
    
#initiation of ancillary to |->
qc.x(anc)
qc.h(anc)
qc.barrier()

#Index 0 -> 3 (both qbits are controlled)
qc.x(qrd)
qc.mcx([0,1],2, mode='noancilla')
qc.x(qrd)

qc.x(qrd)
qc.mcx([0,1],3, mode='noancilla')
qc.x(qrd)
qc.barrier()

#Index 1 -> 2 (only qbit 2 is controlled)
qc.x(1)
qc.mcx([0,1],3, mode='noancilla')
qc.x(1)
qc.barrier()

#Index 2 -> 0 (no qbit is controlled)

#Index 3 -> 1 (only qbit 1 is controlled)
qc.mcx([0,1],2, mode='noancilla')
qc.barrier()

#This means that we have a database looking like [3,2,0,1]
#We are looking for lets say number 2 (so index 1)
#Oracle

qc.x([2])
qc.mcx([2,3],4, mode='noancilla')
qc.x([2])
qc.barrier()

#REPETITION TO MAKE SYMMETRY
#Index 3 -> 1 (only qbit 1 is controlled)
qc.mcx([0,1],2, mode='noancilla')
qc.barrier()

#Index 1 -> 2 (only qbit 2 is controlled)
qc.x(1)
qc.mcx([0,1],3, mode='noancilla')
qc.x(1)
qc.barrier()

#Index 0 -> 3 (both qbits are controlled)
qc.x(qrd)
qc.mcx([0,1],3, mode='noancilla')
qc.x(qrd)

qc.x(qrd)
qc.mcx([0,1],2, mode='noancilla')
qc.x(qrd)
qc.barrier()

#DIFFUSION OP
qc.h(range(n))
qc.x(range(n-1))

#It is difficult to do an actual multi-control-Z gate, so instead we make a MCX surrounded by H gates to transform it into a MCZ
qc.x(1)
qc.h(1)
qc.mct([0], 1) 
qc.h(1)
qc.x(1)

#The other symmetric side of the diffusor
qc.x(range(n-1))
qc.h(range(n))
qc.barrier()


qc.barrier()
qc.measure(range(2),cr)

qc.draw(fold=-1)

┌───┐      ░ ┌───┐     ┌───┐┌───┐     ┌───┐ ░                 ░       ░                 ░       ░                 ░ ┌───┐     ┌───┐┌───┐     ┌───┐ ░ ┌───┐┌───┐          ┌───┐┌───┐      ░  ░ ┌─┐   
      d_0: ┤ H ├──────░─┤ X ├──■──┤ X ├┤ X ├──■──┤ X ├─░────────■────────░───■───░─────────────────░───■───░────────■────────░─┤ X ├──■──┤ X ├┤ X ├──■──┤ X ├─░─┤ H ├┤ X ├───────■──┤ X ├┤ H ├──────░──░─┤M├───
           ├───┤      ░ ├───┤  │  ├───┤├───┤  │  ├───┤ ░ ┌───┐  │  ┌───┐ ░   │   ░                 ░   │   ░ ┌───┐  │  ┌───┐ ░ ├───┤  │  ├───┤├───┤  │  ├───┤ ░ ├───┤├───┤┌───┐┌─┴─┐├───┤├───┤┌───┐ ░  ░ └╥┘┌─┐
      d_1: ┤ H ├──────░─┤ X ├──■──┤ X ├┤ X ├──■──┤ X ├─░─┤ X ├──■──┤ X ├─░───■───░─────────────────░───■───░─┤ X ├──■──┤ X ├─░─┤ X ├──■──┤ X ├┤ X ├──■──┤ X ├─░─┤ H ├┤ X ├┤ H ├┤ X ├┤ H ├┤ X ├┤ H ├─░──░──╫─┤M├
           └───┘      ░ └───┘┌─┴─┐└───┘└───┘  │  └───┘ ░ └───┘  │  └───┘ ░ ┌─┴─┐ ░ ┌───┐     ┌───┐ ░ ┌─┴─┐ ░ └───┘  │  └───┘ ░ └───┘  │  └───┘└───┘┌─┴─┐└───┘ ░ └───┘└───┘└───┘└───┘└───┘└───┘└───┘ ░  ░  ║ └╥┘
      t_0: ───────────░──────┤ X ├────────────┼────────░────────┼────────░─┤ X ├─░─┤ X ├──■──┤ X ├─░─┤ X ├─░────────┼────────░────────┼────────────┤ X ├──────░─────────────────────────────────────░──░──╫──╫─
                      ░      └───┘          ┌─┴─┐      ░      ┌─┴─┐      ░ └───┘ ░ └───┘  │  └───┘ ░ └───┘ ░      ┌─┴─┐      ░      ┌─┴─┐          └───┘      ░                                     ░  ░  ║  ║ 
      t_1: ───────────░─────────────────────┤ X ├──────░──────┤ X ├──────░───────░────────■────────░───────░──────┤ X ├──────░──────┤ X ├─────────────────────░─────────────────────────────────────░──░──╫──╫─
           ┌───┐┌───┐ ░                     └───┘      ░      └───┘      ░       ░      ┌─┴─┐      ░       ░      └───┘      ░      └───┘                     ░                                     ░  ░  ║  ║ 
ancilla_0: ┤ X ├┤ H ├─░────────────────────────────────░─────────────────░───────░──────┤ X ├──────░───────░─────────────────░────────────────────────────────░─────────────────────────────────────░──░──╫──╫─
           └───┘└───┘ ░                                ░                 ░       ░      └───┘      ░       ░                 ░                                ░                                     ░  ░  ║  ║ 
      c: 2/═══════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════╩══╩═
                                                                                                                                                                                                          0  1

In [93]:
############# CIRCUIT SIMULATION #################
aer_sim = Aer.get_backend('aer_simulator')
aer_sim.set_options(device='CPU')
transpiled_qc = transpile(qc, aer_sim)

shots = 2**(n+7) #we repeat the simulation 2048 times
job = aer_sim.run(transpiled_qc, shots=shots)

############# CODE FOR PLOTTING ####################
hist = job.result().get_counts() #simulation output
sort_hist = sorted(hist.items()) #So that the plotting puts everything in the same increasing order of basis
n_hist = {k:v for k,v in sort_hist}

results = {'val':n_hist.keys(),'count':n_hist.values()} #change the formatting of the data to match plotly

fig = px.bar(results, x="val", y="count", text="count")
fig.update_traces(texttemplate='%{text:.2s}', textposition='outside')
fig.update_layout(uniformtext_minsize=6, uniformtext_mode='show')
fig.show()

measured_str = max(hist, key=hist.get)
measured_int = int(measured_str,2)
print(f"The searched value is at index: {measured_int}")

The searched value is at index: 1


## 3,4 registers building (hand encoded), with data array [3,4,9,13]

In [98]:
# Making the circuit in the paper

n = 3 #nb of index qbits
m = 4 #nb of value qbits

qrd = QuantumRegister(n, 'd') #Creates the data registry
qrt = QuantumRegister(m, 't') #Creates the template registry
anc = QuantumRegister(1, 'ancilla') #Creates the ancillary registry for the oracle
cr = ClassicalRegister(n, 'c') #Creates the classical bit measurment output

qc = QuantumCircuit(qrd, qrt, anc, cr) #Makes the circuit with these qbits as input

####### State initialization #########
qc.h(range(n))
    
#initiation of ancillary to |->
qc.x(anc)
qc.h(anc)
qc.barrier()

#Index 0 -> 3 (both qbits are controlled)
qc.x(qrd)
qc.mcx([0,1,2],3, mode='noancilla')
qc.x(qrd)

qc.x(qrd)
qc.mcx([0,1,2],4, mode='noancilla')
qc.x(qrd)
qc.barrier()

#Index 1 -> 4 (only qbit 2 is controlled)
qc.x([1,2])
qc.mcx([0,1,2],5, mode='noancilla')
qc.x([1,2])
qc.barrier()

#Index 2 -> 9 (no qbit is controlled)
qc.x([0,2])
qc.mcx([0,1,2],3, mode='noancilla')
qc.x([0,2])

qc.x([0,2])
qc.mcx([0,1,2],6, mode='noancilla')
qc.x([0,2])
qc.barrier()

#Index 3 -> 13 (only qbit 1 is controlled)
qc.x([2])
qc.mcx([0,1,2],3, mode='noancilla')
qc.x([2])

qc.x([2])
qc.mcx([0,1,2],5, mode='noancilla')
qc.x([2])

qc.x([2])
qc.mcx([0,1,2],6, mode='noancilla')
qc.x([2])
qc.barrier()

#Oracle
qc.x([n+0,n+1,n+3])
qc.mcx(list(range(n,n+m)),anc, mode='noancilla')
qc.x([n+0,n+1,n+3])
qc.barrier()


#REPETITION TO MAKE SYMMETRY
#Index 3 -> 13 (only qbit 1 is controlled)
qc.x([2])
qc.mcx([0,1,2],6, mode='noancilla')
qc.x([2])

qc.x([2])
qc.mcx([0,1,2],5, mode='noancilla')
qc.x([2])

qc.x([2])
qc.mcx([0,1,2],3, mode='noancilla')
qc.x([2])
qc.barrier()

#Index 2 -> 9 (no qbit is controlled)
qc.x([0,2])
qc.mcx([0,1,2],6, mode='noancilla')
qc.x([0,2])

qc.x([0,2])
qc.mcx([0,1,2],3, mode='noancilla')
qc.x([0,2])
qc.barrier()

#Index 1 -> 4 (only qbit 2 is controlled)
qc.x([1,2])
qc.mcx([0,1,2],5, mode='noancilla')
qc.x([1,2])
qc.barrier()

#Index 0 -> 3 (both qbits are controlled)
qc.x(qrd)
qc.mcx([0,1,2],4, mode='noancilla')
qc.x(qrd)

qc.x(qrd)
qc.mcx([0,1,2],3, mode='noancilla')
qc.x(qrd)
qc.barrier()

#DIFFUSION OP
qc.h(range(n))
qc.x(range(n-1))

#It is difficult to do an actual multi-control-Z gate, so instead we make a MCX surrounded by H gates to transform it into a MCZ
qc.x(n-1)
qc.h(n-1)
qc.mct(list(range(n-1)), n-1) 
qc.h(n-1)
qc.x(n-1)

#The other symmetric side of the diffusor
qc.x(range(n-1))
qc.h(range(n))
qc.barrier()


qc.barrier()
qc.measure(range(n),cr)

qc.draw(fold=-1)

┌───┐      ░ ┌───┐     ┌───┐┌───┐     ┌───┐ ░                 ░ ┌───┐     ┌───┐┌───┐     ┌───┐ ░                                               ░                 ░                                               ░ ┌───┐     ┌───┐┌───┐     ┌───┐ ░                 ░ ┌───┐     ┌───┐┌───┐     ┌───┐ ░ ┌───┐┌───┐          ┌───┐┌───┐      ░  ░ ┌─┐      
      d_0: ┤ H ├──────░─┤ X ├──■──┤ X ├┤ X ├──■──┤ X ├─░────────■────────░─┤ X ├──■──┤ X ├┤ X ├──■──┤ X ├─░────────■──────────────■──────────────■────────░─────────────────░────────■──────────────■──────────────■────────░─┤ X ├──■──┤ X ├┤ X ├──■──┤ X ├─░────────■────────░─┤ X ├──■──┤ X ├┤ X ├──■──┤ X ├─░─┤ H ├┤ X ├───────■──┤ X ├┤ H ├──────░──░─┤M├──────
           ├───┤      ░ ├───┤  │  ├───┤├───┤  │  ├───┤ ░ ┌───┐  │  ┌───┐ ░ └───┘  │  └───┘└───┘  │  └───┘ ░        │              │              │        ░                 ░        │              │              │        ░ └───┘  │  └───┘└───┘  │  └───┘ ░ ┌───┐  │  ┌───┐ ░ ├───┤  │  ├───┤├───┤  │  ├───┤ ░ ├───┤├───┤       │  ├───┤├───┤      ░  ░ └╥┘┌─┐   
      d_1: ┤ H ├──────░─┤ X ├──■──┤ X ├┤ X ├──■──┤ X ├─░─┤ X ├──■──┤ X ├─░────────■──────────────■────────░────────■──────────────■──────────────■────────░─────────────────░────────■──────────────■──────────────■────────░────────■──────────────■────────░─┤ X ├──■──┤ X ├─░─┤ X ├──■──┤ X ├┤ X ├──■──┤ X ├─░─┤ H ├┤ X ├───────■──┤ X ├┤ H ├──────░──░──╫─┤M├───
           ├───┤      ░ ├───┤  │  ├───┤├───┤  │  ├───┤ ░ ├───┤  │  ├───┤ ░ ┌───┐  │  ┌───┐┌───┐  │  ┌───┐ ░ ┌───┐  │  ┌───┐┌───┐  │  ┌───┐┌───┐  │  ┌───┐ ░                 ░ ┌───┐  │  ┌───┐┌───┐  │  ┌───┐┌───┐  │  ┌───┐ ░ ┌───┐  │  ┌───┐┌───┐  │  ┌───┐ ░ ├───┤  │  ├───┤ ░ ├───┤  │  ├───┤├───┤  │  ├───┤ ░ ├───┤├───┤┌───┐┌─┴─┐├───┤├───┤┌───┐ ░  ░  ║ └╥┘┌─┐
      d_2: ┤ H ├──────░─┤ X ├──■──┤ X ├┤ X ├──■──┤ X ├─░─┤ X ├──■──┤ X ├─░─┤ X ├──■──┤ X ├┤ X ├──■──┤ X ├─░─┤ X ├──■──┤ X ├┤ X ├──■──┤ X ├┤ X ├──■──┤ X ├─░─────────────────░─┤ X ├──■──┤ X ├┤ X ├──■──┤ X ├┤ X ├──■──┤ X ├─░─┤ X ├──■──┤ X ├┤ X ├──■──┤ X ├─░─┤ X ├──■──┤ X ├─░─┤ X ├──■──┤ X ├┤ X ├──■──┤ X ├─░─┤ H ├┤ X ├┤ H ├┤ X ├┤ H ├┤ X ├┤ H ├─░──░──╫──╫─┤M├
           └───┘      ░ └───┘┌─┴─┐└───┘└───┘  │  └───┘ ░ └───┘  │  └───┘ ░ └───┘┌─┴─┐└───┘└───┘  │  └───┘ ░ └───┘┌─┴─┐└───┘└───┘  │  └───┘└───┘  │  └───┘ ░ ┌───┐     ┌───┐ ░ └───┘  │  └───┘└───┘  │  └───┘└───┘┌─┴─┐└───┘ ░ └───┘  │  └───┘└───┘┌─┴─┐└───┘ ░ └───┘  │  └───┘ ░ └───┘  │  └───┘└───┘┌─┴─┐└───┘ ░ └───┘└───┘└───┘└───┘└───┘└───┘└───┘ ░  ░  ║  ║ └╥┘
      t_0: ───────────░──────┤ X ├────────────┼────────░────────┼────────░──────┤ X ├────────────┼────────░──────┤ X ├────────────┼──────────────┼────────░─┤ X ├──■──┤ X ├─░────────┼──────────────┼────────────┤ X ├──────░────────┼────────────┤ X ├──────░────────┼────────░────────┼────────────┤ X ├──────░─────────────────────────────────────░──░──╫──╫──╫─
                      ░      └───┘          ┌─┴─┐      ░        │        ░      └───┘            │        ░      └───┘            │              │        ░ ├───┤  │  ├───┤ ░        │              │            └───┘      ░        │            └───┘      ░        │        ░      ┌─┴─┐          └───┘      ░                                     ░  ░  ║  ║  ║ 
      t_1: ───────────░─────────────────────┤ X ├──────░────────┼────────░───────────────────────┼────────░───────────────────────┼──────────────┼────────░─┤ X ├──■──┤ X ├─░────────┼──────────────┼───────────────────────░────────┼───────────────────────░────────┼────────░──────┤ X ├─────────────────────░─────────────────────────────────────░──░──╫──╫──╫─
                      ░                     └───┘      ░      ┌─┴─┐      ░                       │        ░                     ┌─┴─┐            │        ░ └───┘  │  └───┘ ░        │            ┌─┴─┐                     ░        │                       ░      ┌─┴─┐      ░      └───┘                     ░                                     ░  ░  ║  ║  ║ 
      t_2: ───────────░────────────────────────────────░──────┤ X ├──────░──────────

In [100]:
############# CIRCUIT SIMULATION #################
aer_sim = Aer.get_backend('aer_simulator')
aer_sim.set_options(device='CPU')
transpiled_qc = transpile(qc, aer_sim)

shots = 2**(n+7) #we repeat the simulation 2048 times
job = aer_sim.run(transpiled_qc, shots=shots)

############# CODE FOR PLOTTING ####################
hist = job.result().get_counts() #simulation output
sort_hist = sorted(hist.items()) #So that the plotting puts everything in the same increasing order of basis
n_hist = {k:v for k,v in sort_hist}

results = {'val':n_hist.keys(),'count':n_hist.values()} #change the formatting of the data to match plotly

fig = px.bar(results, x="val", y="count", text="count")
fig.update_traces(texttemplate='%{text:.2s}', textposition='outside')
fig.update_layout(uniformtext_minsize=6, uniformtext_mode='show')
fig.show()

measured_str = max(hist, key=hist.get)
measured_int = int(measured_str,2)
print(f"The searched value is at index: {measured_int}")

The searched value is at index: 1


## Automatic bitstring encoding

In [101]:
def create_oracle(v_reg, nb_check):
    qc = QuantumCircuit(v_reg + 1)
    
    a = len(bin(nb_check)[2:]) #This gives the length of binary value to encode 
    pos = 0
    for j in list(bin(nb_check)[2:])[::-1]+(v_reg-a)*['0']: #loops through the reverse ordered list composed of the binary encoding of the value
        if not int(j):
            qc.x(pos)
            
        pos += 1
        
    qc.mct(list(range(v_reg)),v_reg)
    
    pos = 0
    for j in list(bin(nb_check)[2:])[::-1]+(v_reg-a)*['0']: #loops through the reverse ordered list composed of the binary encoding of the value
        if not int(j):
            qc.x(pos)
            
        pos += 1
        
    return qc    

In [102]:
create_oracle(4, 2).draw()

┌───┐     ┌───┐
q_0: ┤ X ├──■──┤ X ├
     └───┘  │  └───┘
q_1: ───────■───────
     ┌───┐  │  ┌───┐
q_2: ┤ X ├──■──┤ X ├
     ├───┤  │  ├───┤
q_3: ┤ X ├──■──┤ X ├
     └───┘┌─┴─┐└───┘
q_4: ─────┤ X ├─────
          └───┘

In [103]:
#This function takes in the number of qbits inside the index register and value register, as well as the array
# that works as database and ecodes those values automatically as gates.

#There are some extra x gates that are useless, and those will be removed in a next code update. They don't affect results.
def encode_qc(i_reg,v_reg,val_array):
    
    qc = QuantumCircuit(i_reg + v_reg)
    
    for step in range(len(val_array)):
        
        x_place = 0 #This variable is used to find where to put the NOT in the CNOT encoding
        a = len(bin(val_array[step])[2:]) #This gives the length of binary value to encode 
        
        for j in list(bin(val_array[step])[2:])[::-1]+(v_reg-a)*['0']: #loops through the reverse ordered list composed of the binary encoding of the value

            ### CONTROL X encoding ###
            if int(j):
                
                ### X encoding ###
                pos = 0
                b = len(bin(step)[2:])
                for i in list(bin(step)[2:])[::-1]+(i_reg-b)*['0']:

                    if not int(i) and int(j):
                        qc.x(pos)
                    pos += 1
                
                qc.mcx(list(range(i_reg)), i_reg + x_place)


                ### X encoding ###
                pos = 0
                b = len(bin(step)[2:])
                for i in list(bin(step)[2:])[::-1]+(i_reg-b)*['0']:

                    if not int(i):
                        qc.x(pos)
                    pos += 1
                    
#                 qc.barrier() #uncomment this is you want a more readable circuit, but then cannot be used as gate


            x_place += 1
    
    return qc

In [104]:
encode_qc(3,4,[3,4,9,13]).draw(fold=-1)

┌───┐     ┌───┐┌───┐     ┌───┐          ┌───┐          ┌───┐┌───┐     ┌───┐                                             
q_0: ┤ X ├──■──┤ X ├┤ X ├──■──┤ X ├───────■──┤ X ├───────■──┤ X ├┤ X ├──■──┤ X ├───────■──────────────■──────────────■───────
     ├───┤  │  ├───┤├───┤  │  ├───┤┌───┐  │  ├───┤       │  └───┘└───┘  │  └───┘       │              │              │       
q_1: ┤ X ├──■──┤ X ├┤ X ├──■──┤ X ├┤ X ├──■──┤ X ├───────■──────────────■──────────────■──────────────■──────────────■───────
     ├───┤  │  ├───┤├───┤  │  ├───┤├───┤  │  ├───┤┌───┐  │  ┌───┐┌───┐  │  ┌───┐┌───┐  │  ┌───┐┌───┐  │  ┌───┐┌───┐  │  ┌───┐
q_2: ┤ X ├──■──┤ X ├┤ X ├──■──┤ X ├┤ X ├──■──┤ X ├┤ X ├──■──┤ X ├┤ X ├──■──┤ X ├┤ X ├──■──┤ X ├┤ X ├──■──┤ X ├┤ X ├──■──┤ X ├
     └───┘┌─┴─┐└───┘└───┘  │  └───┘└───┘  │  └───┘└───┘┌─┴─┐└───┘└───┘  │  └───┘└───┘┌─┴─┐└───┘└───┘  │  └───┘└───┘  │  └───┘
q_3: ─────┤ X ├────────────┼──────────────┼────────────┤ X ├────────────┼────────────┤ X ├────────────┼──────────────┼───────
          └───┘          ┌─┴─┐            │            └───┘            │            └───┘            │              │       
q_4: ────────────────────┤ X ├────────────┼─────────────────────────────┼─────────────────────────────┼──────────────┼───────
                         └───┘          ┌─┴─┐                           │                           ┌─┴─┐            │       
q_5: ───────────────────────────────────┤ X ├───────────────────────────┼───────────────────────────┤ X ├────────────┼───────
                                        └───┘                         ┌─┴─┐                         └───┘          ┌─┴─┐     
q_6: ─────────────────────────────────────────────────────────────────┤ X ├────────────────────────────────────────┤ X ├─────
                                                                      └───┘                                        └───┘

In [105]:
encode_qc(3,4,[3,4,9,13]).reverse_ops().draw(fold=-1)

┌───┐          ┌───┐┌───┐     ┌───┐          ┌───┐          ┌───┐┌───┐     ┌───┐
q_0: ───────■──────────────■──────────────■──┤ X ├───────■──┤ X ├┤ X ├──■──┤ X ├───────■──┤ X ├───────■──┤ X ├┤ X ├──■──┤ X ├
            │              │              │  └───┘       │  └───┘└───┘  │  ├───┤       │  ├───┤┌───┐  │  ├───┤├───┤  │  ├───┤
q_1: ───────■──────────────■──────────────■──────────────■──────────────■──┤ X ├───────■──┤ X ├┤ X ├──■──┤ X ├┤ X ├──■──┤ X ├
     ┌───┐  │  ┌───┐┌───┐  │  ┌───┐┌───┐  │  ┌───┐┌───┐  │  ┌───┐┌───┐  │  ├───┤┌───┐  │  ├───┤├───┤  │  ├───┤├───┤  │  ├───┤
q_2: ┤ X ├──■──┤ X ├┤ X ├──■──┤ X ├┤ X ├──■──┤ X ├┤ X ├──■──┤ X ├┤ X ├──■──┤ X ├┤ X ├──■──┤ X ├┤ X ├──■──┤ X ├┤ X ├──■──┤ X ├
     └───┘  │  └───┘└───┘  │  └───┘└───┘┌─┴─┐└───┘└───┘  │  └───┘└───┘┌─┴─┐└───┘└───┘  │  └───┘└───┘  │  └───┘└───┘┌─┴─┐└───┘
q_3: ───────┼──────────────┼────────────┤ X ├────────────┼────────────┤ X ├────────────┼──────────────┼────────────┤ X ├─────
            │              │            └───┘            │            └───┘            │            ┌─┴─┐          └───┘     
q_4: ───────┼──────────────┼─────────────────────────────┼─────────────────────────────┼────────────┤ X ├────────────────────
            │            ┌─┴─┐                           │                           ┌─┴─┐          └───┘                    
q_5: ───────┼────────────┤ X ├───────────────────────────┼───────────────────────────┤ X ├───────────────────────────────────
          ┌─┴─┐          └───┘                         ┌─┴─┐                         └───┘                                   
q_6: ─────┤ X ├────────────────────────────────────────┤ X ├─────────────────────────────────────────────────────────────────
          └───┘                                        └───┘

In [106]:
data_arr = [3,4,9,13,2,7]
search_val = 2 #This is the value that will be in the oracle and be searched. Value 2 is currently on index 4.

n = 3 #number of index qbits
m = 4 #number of value qbits

qrd = QuantumRegister(n, 'd') #Creates the data registry
qrt = QuantumRegister(m, 't') #Creates the template registry
anc = QuantumRegister(1, 'ancilla') #Creates the ancillary registry for the oracle
cr = ClassicalRegister(n, 'c') #Creates the classical bit measurment output

qc = QuantumCircuit(qrd, qrt, anc, cr) #Makes the circuit with these qbits as input

encode = encode_qc(n,m,data_arr).to_gate()
encode.label = "Encoding database"

orac = create_oracle(m, search_val).to_gate()
orac.label = "Oracle"

####### State initialization #########
qc.h(range(n))
    
#initiation of ancillary to |->
qc.x(anc)
qc.h(anc)
qc.barrier()
######################################

#ENCODING
qc.append(encode, range(n+m))

#ORACLE
qc.append(orac, range(n,n+m+1))

#REVERSE ENCODING
qc.append(encode.reverse_ops(), range(n+m))

#DIFFUSION OP
qc.h(range(n))
qc.x(range(n-1))

#It is difficult to do an actual multi-control-Z gate, so instead we make a MCX surrounded by H gates to transform it into a MCZ
qc.x(n-1)
qc.h(n-1)
qc.mct(list(range(n-1)), n-1) 
qc.h(n-1)
qc.x(n-1)

#The other symmetric side of the diffusor
qc.x(range(n-1))
qc.h(range(n))
qc.barrier()


qc.barrier()
qc.measure(range(n),cr)

qc.draw(fold=-1)

┌───┐      ░ ┌────────────────────┐           ┌────────────────────┐┌───┐┌───┐          ┌───┐┌───┐      ░  ░ ┌─┐      
      d_0: ┤ H ├──────░─┤0                   ├───────────┤0                   ├┤ H ├┤ X ├───────■──┤ X ├┤ H ├──────░──░─┤M├──────
           ├───┤      ░ │                    │           │                    │├───┤├───┤       │  ├───┤├───┤      ░  ░ └╥┘┌─┐   
      d_1: ┤ H ├──────░─┤1                   ├───────────┤1                   ├┤ H ├┤ X ├───────■──┤ X ├┤ H ├──────░──░──╫─┤M├───
           ├───┤      ░ │                    │           │                    │├───┤├───┤┌───┐┌─┴─┐├───┤├───┤┌───┐ ░  ░  ║ └╥┘┌─┐
      d_2: ┤ H ├──────░─┤2                   ├───────────┤2                   ├┤ H ├┤ X ├┤ H ├┤ X ├┤ H ├┤ X ├┤ H ├─░──░──╫──╫─┤M├
           └───┘      ░ │                    │┌─────────┐│                    │└───┘└───┘└───┘└───┘└───┘└───┘└───┘ ░  ░  ║  ║ └╥┘
      t_0: ───────────░─┤3 Encoding database ├┤0        ├┤3 Encoding database ├────────────────────────────────────░──░──╫──╫──╫─
                      ░ │                    ││         ││                    │                                    ░  ░  ║  ║  ║ 
      t_1: ───────────░─┤4                   ├┤1        ├┤4                   ├────────────────────────────────────░──░──╫──╫──╫─
                      ░ │                    ││         ││                    │                                    ░  ░  ║  ║  ║ 
      t_2: ───────────░─┤5                   ├┤2 Oracle ├┤5                   ├────────────────────────────────────░──░──╫──╫──╫─
                      ░ │                    ││         ││                    │                                    ░  ░  ║  ║  ║ 
      t_3: ───────────░─┤6                   ├┤3        ├┤6                   ├────────────────────────────────────░──░──╫──╫──╫─
           ┌───┐┌───┐ ░ └────────────────────┘│         │└────────────────────┘                                    ░  ░  ║  ║  ║ 
ancilla_0: ┤ X ├┤ H ├─░───────────────────────┤4        ├──────────────────────────────────────────────────────────░──░──╫──╫──╫─
           └───┘└───┘ ░                       └─────────┘                                                          ░  ░  ║  ║  ║ 
      c: 3/══════════════════════════════════════════════════════════════════════════════════════════════════════════════╩══╩══╩═
                                                                                                                         0  1  2

In [108]:
############# CIRCUIT SIMULATION #################
aer_sim = Aer.get_backend('aer_simulator')
aer_sim.set_options(device='CPU')
transpiled_qc = transpile(qc, aer_sim)

shots = 2**(n+7) #we repeat the simulation 2048 times
job = aer_sim.run(transpiled_qc, shots=shots)

############# CODE FOR PLOTTING ####################
hist = job.result().get_counts() #simulation output
sort_hist = sorted(hist.items()) #So that the plotting puts everything in the same increasing order of basis
n_hist = {k:v for k,v in sort_hist}

results = {'val':n_hist.keys(),'count':n_hist.values()} #change the formatting of the data to match plotly

fig = px.bar(results, x="val", y="count", text="count")
fig.update_traces(texttemplate='%{text:.2s}', textposition='outside')
fig.update_layout(uniformtext_minsize=6, uniformtext_mode='show')
fig.show()

############## CODE FOR LISTING THE RESULTS #####################
trig = []
max_c = max(results['count'])
for i,j in zip(results['val'],results['count']):
    if int(j) > max_c-0.33*max_c: #only consideres values that are higher than one third less
        trig.append({'index_dec':int(i,2),'index_bin':i,'counts':j})

for t in trig:
    print(f"Index {t['index_dec']} holds the value {data_arr[t['index_dec']]}, which is value being looked for.")

Index 4 holds the value 2, which is value being looked for.


# Bit String Comparator

In [109]:
def U_c():
    qc = QuantumCircuit(4)
    
    qc.x(1)
    qc.mcx([0,1],2, mode='noancilla')
    qc.x([0,1])
    qc.mcx([0,1],3, mode='noancilla')
    qc.x(0)
    
    return qc

def U_c_T():
    qc = QuantumCircuit(4)
    
    qc.x(0)
    qc.mcx([0,1],3, mode='noancilla')
    qc.x([0,1])
    qc.mcx([0,1],2, mode='noancilla')
    qc.x(1)
    
    return qc

In [110]:
U_c().draw()

┌───┐     ┌───┐
q_0: ───────■──┤ X ├──■──┤ X ├
     ┌───┐  │  ├───┤  │  └───┘
q_1: ┤ X ├──■──┤ X ├──■───────
     └───┘┌─┴─┐└───┘  │       
q_2: ─────┤ X ├───────┼───────
          └───┘     ┌─┴─┐     
q_3: ───────────────┤ X ├─────
                    └───┘

In [111]:
U_c_T().draw()

┌───┐     ┌───┐          
q_0: ┤ X ├──■──┤ X ├──■───────
     └───┘  │  ├───┤  │  ┌───┐
q_1: ───────■──┤ X ├──■──┤ X ├
            │  └───┘┌─┴─┐└───┘
q_2: ───────┼───────┤ X ├─────
          ┌─┴─┐     └───┘     
q_3: ─────┤ X ├───────────────
          └───┘

In [115]:
# Making the circuit in the paper

len_b = 3 #length of the bitstrings to compare
n = 5*len_b #number of qbits

base = QuantumRegister(n, 'd') #Creates the data registry
cr = ClassicalRegister(2, 'c') #Creates the classical bit measurment output

qc = QuantumCircuit(base, cr) #Makes the circuit with these qbits as input

####### State initialization #########

z = [1, 0] # base vect |0>
o = [0, 1] # base vect |1>
init_st_data = [z,o,o] #data bit stream 011000 (this is what we will compare the templates to and hope it matches)

qc.x(0)
# qc.x(5)
# qc.x(10)

qc.x(1)
qc.x(6)
qc.x(11)

qc.barrier()

u_c = U_c().to_gate()
u_c.label = "U_c"

qc.append(u_c, range(4))
qc.append(u_c, range(5,9))
qc.append(u_c, range(10,14))

qc.barrier()

qc.x(2)
qc.x(3)
qc.mcx([2,3],4, mode='noancilla')
qc.x(2)
qc.x(3)

qc.barrier()

qc.x(7)
qc.x(8)
qc.mcx([7,8],9, mode='noancilla')
qc.x(7)
qc.x(8)

qc.barrier()

qc.mcx([9,12],7, mode='noancilla')
qc.mcx([9,13],8, mode='noancilla')

qc.barrier()

qc.mcx([4,7],2, mode='noancilla')
qc.mcx([4,8],3, mode='noancilla')
    
qc.barrier()

qc.measure([2,3],cr)

qc.draw(fold=-1)

┌───┐ ░ ┌──────┐ ░                 ░                 ░            ░            ░       
 d_0: ┤ X ├─░─┤0     ├─░─────────────────░─────────────────░────────────░────────────░───────
      ├───┤ ░ │      │ ░                 ░                 ░            ░            ░       
 d_1: ┤ X ├─░─┤1     ├─░─────────────────░─────────────────░────────────░────────────░───────
      └───┘ ░ │  U_c │ ░ ┌───┐     ┌───┐ ░                 ░            ░ ┌───┐      ░ ┌─┐   
 d_2: ──────░─┤2     ├─░─┤ X ├──■──┤ X ├─░─────────────────░────────────░─┤ X ├──────░─┤M├───
            ░ │      │ ░ ├───┤  │  ├───┤ ░                 ░            ░ └─┬─┘┌───┐ ░ └╥┘┌─┐
 d_3: ──────░─┤3     ├─░─┤ X ├──■──┤ X ├─░─────────────────░────────────░───┼──┤ X ├─░──╫─┤M├
            ░ └──────┘ ░ └───┘┌─┴─┐└───┘ ░                 ░            ░   │  └─┬─┘ ░  ║ └╥┘
 d_4: ──────░──────────░──────┤ X ├──────░─────────────────░────────────░───■────■───░──╫──╫─
            ░ ┌──────┐ ░      └───┘      ░                 ░            ░   │    │   ░  ║  ║ 
 d_5: ──────░─┤0     ├─░─────────────────░─────────────────░────────────░───┼────┼───░──╫──╫─
      ┌───┐ ░ │      │ ░                 ░                 ░            ░   │    │   ░  ║  ║ 
 d_6: ┤ X ├─░─┤1     ├─░─────────────────░─────────────────░────────────░───┼────┼───░──╫──╫─
      └───┘ ░ │  U_c │ ░                 ░ ┌───┐     ┌───┐ ░ ┌───┐      ░   │    │   ░  ║  ║ 
 d_7: ──────░─┤2     ├─░─────────────────░─┤ X ├──■──┤ X ├─░─┤ X ├──────░───■────┼───░──╫──╫─
            ░ │      │ ░                 ░ ├───┤  │  ├───┤ ░ └─┬─┘┌───┐ ░        │   ░  ║  ║ 
 d_8: ──────░─┤3     ├─░─────────────────░─┤ X ├──■──┤ X ├─░───┼──┤ X ├─░────────■───░──╫──╫─
            ░ └──────┘ ░                 ░ └───┘┌─┴─┐└───┘ ░   │  └─┬─┘ ░            ░  ║  ║ 
 d_9: ──────░──────────░─────────────────░──────┤ X ├──────░───■────■───░────────────░──╫──╫─
            ░ ┌──────┐ ░                 ░      └───┘      ░   │    │   ░            ░  ║  ║ 
d_10: ──────░─┤0     ├─░─────────────────░─────────────────░───┼────┼───░────────────░──╫──╫─
      ┌───┐ ░ │      │ ░                 ░                 ░   │    │   ░            ░  ║  ║ 
d_11: ┤ X ├─░─┤1     ├─░─────────────────░─────────────────░───┼────┼───░────────────░──╫──╫─
      └───┘ ░ │  U_c │ ░                 ░                 ░   │    │   ░            ░  ║  ║ 
d_12: ──────░─┤2     ├─░─────────────────░─────────────────░───■────┼───░────────────░──╫──╫─
            ░ │      │ ░                 ░                 ░        │   ░            ░  ║  ║ 
d_13: ──────░─┤3     ├─░─────────────────░─────────────────░────────■───░────────────░──╫──╫─
            ░ └──────┘ ░                 ░                 ░            ░            ░  ║  ║ 
 c: 2/══════════════════════════════════════════════════════════════════════════════════╩══╩═
                                                                                        0  1

Outpu: $|O_2 O_1>$

In a measurement of the outputs (O1 and O2), 

if O1=1 and O2=0 then a>b; 

if O1=0 and O2=1 then a<b;

if O1=0 and O2=0 then a=b.

In [116]:
############# CIRCUIT SIMULATION #################
aer_sim = Aer.get_backend('aer_simulator')
aer_sim.set_options(device='CPU')
transpiled_qc = transpile(qc, aer_sim)

shots = 2**(len_b+7) #we repeat the simulation 2048 times
job = aer_sim.run(transpiled_qc, shots=shots)

############# CODE FOR PLOTTING ####################
hist = job.result().get_counts() #simulation output
sort_hist = sorted(hist.items()) #So that the plotting puts everything in the same increasing order of basis
n_hist = {k:v for k,v in sort_hist}

results = {'val':n_hist.keys(),'count':n_hist.values()} #change the formatting of the data to match plotly

fig = px.bar(results, x="val", y="count", text="count")
fig.update_traces(texttemplate='%{text:.2s}', textposition='outside')
fig.update_layout(uniformtext_minsize=6, uniformtext_mode='show')
fig.show()

# Bit String Comparator inside Grover

In [117]:
# Making the circuit in the paper

len_b = 3 #length of the bitstrings to compare
n = 5*len_b #number of qbits


base = QuantumRegister(n, 'd') #Creates the data registry
cr = ClassicalRegister(len_b, 'c') #Creates the classical bit measurment output

qc = QuantumCircuit(base, cr) #Makes the circuit with these qbits as input

####### State initialization #########

qc.x(14)
qc.h(14)

qc.x(0) #first bit (2^2)
# qc.x(5) #second bit (2^1)
# qc.x(10) #third bit (2^0)

qc.h(1) #first superp. bit
qc.h(6) #second superp. bit
qc.h(11) #third superp. bit

u_c = U_c().to_gate()
u_c.label = "U_c"

u_c_T = U_c_T().to_gate()
u_c_T.label = "U_c†"

#################### GROVER CIRCUIT ######################

qc.append(u_c, range(4))
qc.append(u_c, range(5,9))
qc.append(u_c, range(10,14))

qc.barrier()

qc.x(2)
qc.x(3)
qc.mcx([2,3],4, mode='noancilla')
qc.x(2)
qc.x(3)

qc.x(7)
qc.x(8)
qc.mcx([7,8],9, mode='noancilla')
qc.x(7)
qc.x(8)

qc.barrier()

qc.mcx([9,12],7, mode='noancilla')
qc.mcx([9,13],8, mode='noancilla')

qc.barrier()

qc.mcx([4,7],2, mode='noancilla')
qc.mcx([4,8],3, mode='noancilla')

qc.barrier()

qc.cx(3,14)

qc.barrier()

qc.mcx([4,8],3, mode='noancilla')
qc.mcx([4,7],2, mode='noancilla')

qc.barrier()

qc.mcx([9,13],8, mode='noancilla')
qc.mcx([9,12],7, mode='noancilla')

qc.barrier()

qc.x(8)
qc.x(7)
qc.mcx([7,8],9, mode='noancilla')
qc.x(8)
qc.x(7)


qc.x(3)
qc.x(2)
qc.mcx([2,3],4, mode='noancilla')
qc.x(3)
qc.x(2)

qc.barrier()

qc.append(u_c_T, range(10,14))
qc.append(u_c_T, range(5,9))
qc.append(u_c_T, range(4))


qc.barrier()


#DIFFUSION
qc.h([1,6,11])#range(n-1))
qc.x([1,6,11])#range(n-1))

#It is difficult to do an actual multi-control-Z gate, so instead we make a MCX surrounded by H gates to transform it into a MCZ
qc.h(11)#n-2)
qc.mct([1,6],11)#list(range(n-2)), n-2) 
qc.h(11)#n-2)

#The other symmetric side of the diffusor
qc.x([1,6,11])#range(n-1))
qc.h([1,6,11])#range(n-1))


qc.measure([11,6,1],cr)

qc.draw(fold=-1)

┌───┐┌──────┐ ░                 ░            ░            ░       ░            ░            ░                 ░ ┌───────┐ ░                                             
 d_0: ┤ X ├┤0     ├─░─────────────────░────────────░────────────░───────░────────────░────────────░─────────────────░─┤0      ├─░─────────────────────────────────────────────
      ├───┤│      │ ░                 ░            ░            ░       ░            ░            ░                 ░ │       │ ░ ┌───┐┌───┐          ┌───┐┌───┐        ┌─┐   
 d_1: ┤ H ├┤1     ├─░─────────────────░────────────░────────────░───────░────────────░────────────░─────────────────░─┤1      ├─░─┤ H ├┤ X ├───────■──┤ X ├┤ H ├────────┤M├───
      └───┘│  U_c │ ░ ┌───┐     ┌───┐ ░            ░ ┌───┐      ░       ░      ┌───┐ ░            ░ ┌───┐     ┌───┐ ░ │  U_c† │ ░ └───┘└───┘       │  └───┘└───┘        └╥┘   
 d_2: ─────┤2     ├─░─┤ X ├──■──┤ X ├─░────────────░─┤ X ├──────░───────░──────┤ X ├─░────────────░─┤ X ├──■──┤ X ├─░─┤2      ├─░──────────────────┼─────────────────────╫────
           │      │ ░ ├───┤  │  ├───┤ ░            ░ └─┬─┘┌───┐ ░       ░ ┌───┐└─┬─┘ ░            ░ ├───┤  │  ├───┤ ░ │       │ ░                  │                     ║    
 d_3: ─────┤3     ├─░─┤ X ├──■──┤ X ├─░────────────░───┼──┤ X ├─░───■───░─┤ X ├──┼───░────────────░─┤ X ├──■──┤ X ├─░─┤3      ├─░──────────────────┼─────────────────────╫────
           └──────┘ ░ └───┘┌─┴─┐└───┘ ░            ░   │  └─┬─┘ ░   │   ░ └─┬─┘  │   ░            ░ └───┘┌─┴─┐└───┘ ░ └───────┘ ░                  │                     ║    
 d_4: ──────────────░──────┤ X ├──────░────────────░───■────■───░───┼───░───■────■───░────────────░──────┤ X ├──────░───────────░──────────────────┼─────────────────────╫────
           ┌──────┐ ░      └───┘      ░            ░   │    │   ░   │   ░   │    │   ░            ░      └───┘      ░ ┌───────┐ ░                  │                     ║    
 d_5: ─────┤0     ├─░─────────────────░────────────░───┼────┼───░───┼───░───┼────┼───░────────────░─────────────────░─┤0      ├─░──────────────────┼─────────────────────╫────
      ┌───┐│      │ ░                 ░            ░   │    │   ░   │   ░   │    │   ░            ░                 ░ │       │ ░ ┌───┐┌───┐       │  ┌───┐┌───┐     ┌─┐ ║    
 d_6: ┤ H ├┤1     ├─░─────────────────░────────────░───┼────┼───░───┼───░───┼────┼───░────────────░─────────────────░─┤1      ├─░─┤ H ├┤ X ├───────■──┤ X ├┤ H ├─────┤M├─╫────
      └───┘│  U_c │ ░ ┌───┐     ┌───┐ ░ ┌───┐      ░   │    │   ░   │   ░   │    │   ░      ┌───┐ ░ ┌───┐     ┌───┐ ░ │  U_c† │ ░ └───┘└───┘       │  └───┘└───┘     └╥┘ ║    
 d_7: ─────┤2     ├─░─┤ X ├──■──┤ X ├─░─┤ X ├──────░───■────┼───░───┼───░───┼────■───░──────┤ X ├─░─┤ X ├──■──┤ X ├─░─┤2      ├─░──────────────────┼──────────────────╫──╫────
           │      │ ░ ├───┤  │  ├───┤ ░ └─┬─┘┌───┐ ░        │   ░   │   ░   │        ░ ┌───┐└─┬─┘ ░ ├───┤  │  ├───┤ ░ │       │ ░                  │                  ║  ║    
 d_8: ─────┤3     ├─░─┤ X ├──■──┤ X ├─░───┼──┤ X ├─░────────■───░───┼───░───■────────░─┤ X ├──┼───░─┤ X ├──■──┤ X ├─░─┤3      ├─░──────────────────┼──────────────────╫──╫────
           └──────┘ ░ └───┘┌─┴─┐└───┘ ░   │  └─┬─┘ ░            ░   │   ░            ░ └─┬─┘  │   ░ └───┘┌─┴─┐└───┘ ░ └───────┘ ░                  │                  ║  ║    
 d_9: ──────────────░──────┤ X ├──────░───■────■───░────────────░───┼───░────────────░───■────■───░──────┤ X ├──────░───────────░──────────────────┼──────────────────╫──╫────
           ┌──────┐ ░      └───┘      ░   │    │   ░            ░   │   ░            ░   │    │   ░      └───┘      ░ ┌───────┐ ░                  │                  ║  ║    
d_10: ─────┤0     ├─░─────────────────░───┼────┼───░────────────░───┼───░────────────░───┼────┼───░─────────────────░─┤0      ├─░──────────────────┼──────────────────╫──╫────
      ┌───┐│      │ ░                 ░   │    │   ░            ░   │   ░            ░   │    │   ░                 ░ │       │ ░ ┌───┐┌───┐┌───┐┌─┴─┐┌───┐┌

Using a = 100 (4 in dec), I should have an output with 

* 101 (5 in dec)
* 011 (6 in dec)
* 111 (7 in dec)

and no

* 000
* 001
* 010
* 110
* 100



In [123]:
############# CIRCUIT SIMULATION #################
aer_sim = Aer.get_backend('aer_simulator')
aer_sim.set_options(device='CPU')
transpiled_qc = transpile(qc, aer_sim)

shots = 2**(len_b+7) #we repeat the simulation 2048 times
job = aer_sim.run(transpiled_qc, shots=shots)

############# CODE FOR PLOTTING ####################
hist = job.result().get_counts() #simulation output
sort_hist = sorted(hist.items()) #So that the plotting puts everything in the same increasing order of basis
n_hist = {k:v for k,v in sort_hist}

results = {'val':n_hist.keys(),'count':n_hist.values()} #change the formatting of the data to match plotly

fig = px.bar(results, x="val", y="count", text="count")
fig.update_traces(texttemplate='%{text:.2s}', textposition='outside')
fig.update_layout(uniformtext_minsize=6, uniformtext_mode='show')
fig.show()

############## CODE FOR LISTING THE RESULTS #####################
trig = []
max_c = max(results['count'])
for i,j in zip(results['val'],results['count']):
    if int(j) > max_c-0.33*max_c: #only consideres values that are higher than one third less
        trig.append({'index_dec':int(i,2),'index_bin':i,'counts':j})

for t in trig:
    print(f"Output, binary:  {t['index_bin']}; decimal: {t['index_dec']}")

Output, binary:  101; decimal: 5
Output, binary:  110; decimal: 6
Output, binary:  111; decimal: 7


## Creating functions to make the circuit scalable to different sized bitstrings 

In [129]:
def n_comp_p1(n):
    qc = QuantumCircuit(5*n-1)
    u_c = U_c().to_gate()
    u_c.label = "U_c"
    
    for i in range(n):
        qc.append(u_c, range(i*5,i*5 + 4))
        
    for i in range(n-1):
        qc.x(i*5 + 2)
        qc.x(i*5 + 3)
        qc.mcx([i*5 + 2, i*5 + 3], i*5 + 4, mode='noancilla')
        qc.x(i*5 + 2)
        qc.x(i*5 + 3)
        
    for i in reversed(range(n-1)):
        qc.mcx([i*5 + 4,i*5 + 7], i*5 + 2, mode='noancilla')
        qc.mcx([i*5 + 4,i*5 + 8], i*5 + 3, mode='noancilla')
        
    return qc

In [130]:
print(n_comp_p1(4).draw())

      ┌──────┐                                   
 q_0: ┤0     ├───────────────────────────────────
      │      │                                   
 q_1: ┤1     ├───────────────────────────────────
      │  U_c │┌───┐     ┌───┐          ┌───┐     
 q_2: ┤2     ├┤ X ├──■──┤ X ├──────────┤ X ├─────
      │      │├───┤  │  ├───┤          └─┬─┘┌───┐
 q_3: ┤3     ├┤ X ├──■──┤ X ├────────────┼──┤ X ├
      └──────┘└───┘┌─┴─┐└───┘            │  └─┬─┘
 q_4: ─────────────┤ X ├─────────────────■────■──
      ┌──────┐     └───┘                 │    │  
 q_5: ┤0     ├───────────────────────────┼────┼──
      │      │                           │    │  
 q_6: ┤1     ├───────────────────────────┼────┼──
      │  U_c │┌───┐     ┌───┐     ┌───┐  │    │  
 q_7: ┤2     ├┤ X ├──■──┤ X ├─────┤ X ├──■────┼──
      │      │├───┤  │  ├───┤     └─┬─┘┌───┐  │  
 q_8: ┤3     ├┤ X ├──■──┤ X ├───────┼──┤ X ├──■──
      └──────┘└───┘┌─┴─┐└───┘       │  └─┬─┘     
 q_9: ─────────────┤ X ├────────────■────■───────


In [150]:
def n_comparator(n):
    qc = QuantumCircuit(5*n)
    
###############################################
# Inverse the commenting here if you want to see inside the circuit blocks

    c1 = n_comp_p1(n).to_gate()
    c1.label = "Comp"
    qc.append(c1, range(5*n-1))
    
#     c1 = n_comp_p1(n)
#     qc.compose(c1, range(5*n-1),inplace=True)

###############################################

#     qc.barrier()
    qc.cx(3,5*n-1)
#     qc.barrier()
    
###############################################
# Inverse the commenting here if you want to see inside the circuit blocks

    c2 = c1.reverse_ops()
    c2.label = "Comp_R"
    qc.append(c2, range(5*n-1))

#     c2 = c1.reverse_ops()
#     qc.compose(c2, range(5*n-1),inplace=True)

###############################################
    
    return qc

In [151]:
n_comparator(3).draw(fold=-1)

┌────────┐     ┌──────────┐
 q_0: ┤0       ├─────┤0         ├
      │        │     │          │
 q_1: ┤1       ├─────┤1         ├
      │        │     │          │
 q_2: ┤2       ├─────┤2         ├
      │        │     │          │
 q_3: ┤3       ├──■──┤3         ├
      │        │  │  │          │
 q_4: ┤4       ├──┼──┤4         ├
      │        │  │  │          │
 q_5: ┤5       ├──┼──┤5         ├
      │        │  │  │          │
 q_6: ┤6       ├──┼──┤6         ├
      │   Comp │  │  │   Comp_R │
 q_7: ┤7       ├──┼──┤7         ├
      │        │  │  │          │
 q_8: ┤8       ├──┼──┤8         ├
      │        │  │  │          │
 q_9: ┤9       ├──┼──┤9         ├
      │        │  │  │          │
q_10: ┤10      ├──┼──┤10        ├
      │        │  │  │          │
q_11: ┤11      ├──┼──┤11        ├
      │        │  │  │          │
q_12: ┤12      ├──┼──┤12        ├
      │        │  │  │          │
q_13: ┤13      ├──┼──┤13        ├
      └────────┘┌─┴─┐└──────────┘
q_14: ──────────┤ X ├────────────
                └───┘

### Here we build the circuit that will adapt depending on the size of the bitstrings to compare

In [135]:
# Making the circuit in the paper

len_b = 4 #length of the bitstrings to compare
n = 5*len_b #number of qbits

base = QuantumRegister(n, 'd') #Creates the data registry
cr = ClassicalRegister(len_b, 'c') #Creates the classical bit measurment output

qc = QuantumCircuit(base, cr) #Makes the circuit with these qbits as input

####### State initialization #########

qc.x(n-1)
qc.h(n-1)

qc.x(0) #first bit (2^3)
# qc.x(5) #second bit (2^2)
qc.x(10) #third bit (2^1)
qc.x(15) #third bit (2^0)

qc.h(1) #first superp. bit
qc.h(6) #second superp. bit
qc.h(11) #third superp. bit
qc.h(16) #third superp. bit

#################### GROVER CIRCUIT ######################
comp_qc = n_comparator(len_b).to_gate()
comp_qc.label = "Comparator"

qc.append(comp_qc, range(n))


#DIFFUSION
qc.h([1,6,11,16])#range(n-1))
qc.x([1,6,11,16])#range(n-1))

#It is difficult to do an actual multi-control-Z gate, so instead we make a MCX surrounded by H gates to transform it into a MCZ
qc.h(16)
qc.mct([1,6,11],16)#list(range(n-2)), n-2) 
qc.h(16)

#The other symmetric side of the diffusor
qc.x([1,6,11,16])#range(n-1))
qc.h([1,6,11,16])#range(n-1))


qc.measure([16,11,6,1],cr)

qc.draw(fold=-1)

┌───┐     ┌──────────────┐                                               
 d_0: ┤ X ├─────┤0             ├───────────────────────────────────────────────
      ├───┤     │              │┌───┐┌───┐          ┌───┐┌───┐           ┌─┐   
 d_1: ┤ H ├─────┤1             ├┤ H ├┤ X ├───────■──┤ X ├┤ H ├───────────┤M├───
      └───┘     │              │└───┘└───┘       │  └───┘└───┘           └╥┘   
 d_2: ──────────┤2             ├─────────────────┼────────────────────────╫────
                │              │                 │                        ║    
 d_3: ──────────┤3             ├─────────────────┼────────────────────────╫────
                │              │                 │                        ║    
 d_4: ──────────┤4             ├─────────────────┼────────────────────────╫────
                │              │                 │                        ║    
 d_5: ──────────┤5             ├─────────────────┼────────────────────────╫────
      ┌───┐     │              │┌───┐┌───┐       │  ┌───┐┌───┐        ┌─┐ ║    
 d_6: ┤ H ├─────┤6             ├┤ H ├┤ X ├───────■──┤ X ├┤ H ├────────┤M├─╫────
      └───┘     │              │└───┘└───┘       │  └───┘└───┘        └╥┘ ║    
 d_7: ──────────┤7             ├─────────────────┼─────────────────────╫──╫────
                │              │                 │                     ║  ║    
 d_8: ──────────┤8             ├─────────────────┼─────────────────────╫──╫────
                │              │                 │                     ║  ║    
 d_9: ──────────┤9             ├─────────────────┼─────────────────────╫──╫────
      ┌───┐     │   Comparator │                 │                     ║  ║    
d_10: ┤ X ├─────┤10            ├─────────────────┼─────────────────────╫──╫────
      ├───┤     │              │┌───┐┌───┐       │  ┌───┐┌───┐     ┌─┐ ║  ║    
d_11: ┤ H ├─────┤11            ├┤ H ├┤ X ├───────■──┤ X ├┤ H ├─────┤M├─╫──╫────
      └───┘     │              │└───┘└───┘       │  └───┘└───┘     └╥┘ ║  ║    
d_12: ──────────┤12            ├─────────────────┼──────────────────╫──╫──╫────
                │              │                 │                  ║  ║  ║    
d_13: ──────────┤13            ├─────────────────┼──────────────────╫──╫──╫────
                │              │                 │                  ║  ║  ║    
d_14: ──────────┤14            ├─────────────────┼──────────────────╫──╫──╫────
      ┌───┐     │              │                 │                  ║  ║  ║    
d_15: ┤ X ├─────┤15            ├─────────────────┼──────────────────╫──╫──╫────
      ├───┤     │              │┌───┐┌───┐┌───┐┌─┴─┐┌───┐┌───┐┌───┐ ║  ║  ║ ┌─┐
d_16: ┤ H ├─────┤16            ├┤ H ├┤ X ├┤ H ├┤ X ├┤ H ├┤ X ├┤ H ├─╫──╫──╫─┤M├
      └───┘     │              │└───┘└───┘└───┘└───┘└───┘└───┘└───┘ ║  ║  ║ └╥┘
d_17: ──────────┤17            ├────────────────────────────────────╫──╫──╫──╫─
                │              │                                    ║  ║  ║  ║ 
d_18: ──────────┤18            ├────────────────────────────────────╫──╫──╫──╫─
      ┌───┐┌───┐│              │                                    ║  ║  ║  ║ 
d_19: ┤ X ├┤ H ├┤19            ├────────────────────────────────────╫──╫──╫──╫─
      └───┘└───┘└──────────────┘                                    ║  ║  ║  ║ 
 c: 4/══════════════════════════════════════════════════════════════╩══╩══╩══╩═
                                                                    1  2  3  0

We encoded 1011 (11 in decimal), so we should end with 

* 1100 (12)
* 1101 (13)
* 1110 (14)
* 1111 (15)

In [136]:
############# CIRCUIT SIMULATION #################
aer_sim = Aer.get_backend('aer_simulator')
aer_sim.set_options(device='CPU')
transpiled_qc = transpile(qc, aer_sim)

shots = 2**(len_b+7) #we repeat the simulation 2048 times
job = aer_sim.run(transpiled_qc, shots=shots)

############# CODE FOR PLOTTING ####################
hist = job.result().get_counts() #simulation output
sort_hist = sorted(hist.items()) #So that the plotting puts everything in the same increasing order of basis
n_hist = {k:v for k,v in sort_hist}

results = {'val':n_hist.keys(),'count':n_hist.values()} #change the formatting of the data to match plotly

fig = px.bar(results, x="val", y="count", text="count")
fig.update_traces(texttemplate='%{text:.2s}', textposition='outside')
fig.update_layout(uniformtext_minsize=6, uniformtext_mode='show')
fig.show()

############## CODE FOR LISTING THE RESULTS #####################
trig = []
max_c = max(results['count'])
for i,j in zip(results['val'],results['count']):
    if int(j) > max_c-0.33*max_c: #only consideres values that are higher than one third less
        trig.append({'index_dec':int(i,2),'index_bin':i,'counts':j})

for t in trig:
    print(f"Output, binary:  {t['index_bin']}; decimal: {t['index_dec']}")

Output, binary:  1100; decimal: 12
Output, binary:  1101; decimal: 13
Output, binary:  1110; decimal: 14
Output, binary:  1111; decimal: 15


which we do! :)

## Making the Grover Iteration into a function to be able to apply it multiple times to lower the prob. of wrong answers (same comparison as two above)

In [7]:
def Grover_iter_QBSC():
    len_b = 3
    n = 5*len_b
    qc = QuantumCircuit(15)
    
    comp_qc = n_comparator(len_b).to_gate()
    comp_qc.label = "Comparator"

    qc.append(comp_qc, range(n))


    #DIFFUSION
    qc.h([1,6,11])#range(n-1))
    qc.x([1,6,11])#range(n-1))

    #It is difficult to do an actual multi-control-Z gate, so instead we make a MCX surrounded by H gates to transform it into a MCZ
    qc.h(11)#n-2)
    qc.mct([1,6],11)#list(range(n-2)), n-2) 
    qc.h(11)#n-2)

    #The other symmetric side of the diffusor
    qc.x([1,6,11])#range(n-1))
    qc.h([1,6,11])#range(n-1))
    
    return qc

In [139]:
# Making the circuit in the paper

len_b = 3 #length of the bitstrings to compare
n = 5*len_b #number of qbits

base = QuantumRegister(n, 'd') #Creates the data registry
cr = ClassicalRegister(len_b, 'c') #Creates the classical bit measurment output

qc = QuantumCircuit(base, cr) #Makes the circuit with these qbits as input

####### State initialization #########

qc.x(14)
qc.h(14)

qc.x(0) #first bit
# qc.x(5) #second bit
# qc.x(10) #third bit

qc.h(1) #first superp. bit
qc.h(6) #second superp. bit
qc.h(11) #third superp. bit

grit = Grover_iter_QBSC().to_gate() #Transforms the circuit into a usable gate
grit.label = "Grover" #Names the gate

M = 3
for i in range(M):
    qc.append(grit,range(n))

qc.measure([11,6,1],cr)

qc.draw(fold=-1)

┌───┐     ┌──────────┐┌──────────┐┌──────────┐         
 d_0: ┤ X ├─────┤0         ├┤0         ├┤0         ├─────────
      ├───┤     │          ││          ││          │      ┌─┐
 d_1: ┤ H ├─────┤1         ├┤1         ├┤1         ├──────┤M├
      └───┘     │          ││          ││          │      └╥┘
 d_2: ──────────┤2         ├┤2         ├┤2         ├───────╫─
                │          ││          ││          │       ║ 
 d_3: ──────────┤3         ├┤3         ├┤3         ├───────╫─
                │          ││          ││          │       ║ 
 d_4: ──────────┤4         ├┤4         ├┤4         ├───────╫─
                │          ││          ││          │       ║ 
 d_5: ──────────┤5         ├┤5         ├┤5         ├───────╫─
      ┌───┐     │          ││          ││          │   ┌─┐ ║ 
 d_6: ┤ H ├─────┤6         ├┤6         ├┤6         ├───┤M├─╫─
      └───┘     │          ││          ││          │   └╥┘ ║ 
 d_7: ──────────┤7  Grover ├┤7  Grover ├┤7  Grover ├────╫──╫─
                │          ││          ││          │    ║  ║ 
 d_8: ──────────┤8         ├┤8         ├┤8         ├────╫──╫─
                │          ││          ││          │    ║  ║ 
 d_9: ──────────┤9         ├┤9         ├┤9         ├────╫──╫─
                │          ││          ││          │    ║  ║ 
d_10: ──────────┤10        ├┤10        ├┤10        ├────╫──╫─
      ┌───┐     │          ││          ││          │┌─┐ ║  ║ 
d_11: ┤ H ├─────┤11        ├┤11        ├┤11        ├┤M├─╫──╫─
      └───┘     │          ││          ││          │└╥┘ ║  ║ 
d_12: ──────────┤12        ├┤12        ├┤12        ├─╫──╫──╫─
                │          ││          ││          │ ║  ║  ║ 
d_13: ──────────┤13        ├┤13        ├┤13        ├─╫──╫──╫─
      ┌───┐┌───┐│          ││          ││          │ ║  ║  ║ 
d_14: ┤ X ├┤ H ├┤14        ├┤14        ├┤14        ├─╫──╫──╫─
      └───┘└───┘└──────────┘└──────────┘└──────────┘ ║  ║  ║ 
 c: 3/═══════════════════════════════════════════════╩══╩══╩═
                                                     0  1  2

In [140]:
############# CIRCUIT SIMULATION #################
aer_sim = Aer.get_backend('aer_simulator')
aer_sim.set_options(device='CPU')
transpiled_qc = transpile(qc, aer_sim)

shots = 2**(len_b+7) #we repeat the simulation 2048 times
job = aer_sim.run(transpiled_qc, shots=shots)

############# CODE FOR PLOTTING ####################
hist = job.result().get_counts() #simulation output
sort_hist = sorted(hist.items()) #So that the plotting puts everything in the same increasing order of basis
n_hist = {k:v for k,v in sort_hist}

results = {'val':n_hist.keys(),'count':n_hist.values()} #change the formatting of the data to match plotly

fig = px.bar(results, x="val", y="count", text="count")
fig.update_traces(texttemplate='%{text:.2s}', textposition='outside')
fig.update_layout(uniformtext_minsize=6, uniformtext_mode='show')
fig.show()

############## CODE FOR LISTING THE RESULTS #####################
trig = []
max_c = max(results['count'])
for i,j in zip(results['val'],results['count']):
    if int(j) > max_c-0.33*max_c: #only consideres values that are higher than one third less
        trig.append({'index_dec':int(i,2),'index_bin':i,'counts':j})

for t in trig:
    print(f"Output, binary:  {t['index_bin']}; decimal: {t['index_dec']}")

Output, binary:  101; decimal: 5
Output, binary:  110; decimal: 6
Output, binary:  111; decimal: 7


# Assembling both comparator and database into one (final code)

In [141]:
#This function takes in the quantum circuit to add the threshold, the size 
def set_thresh(qc, nb_i_bits, nb_v_bits, val):
    pos = 0
    a = len(bin(val)[2:])
    # This time no [::-1] (no reverse order), as the order of encoding is opposite to the system; instead we first add the 0s and then the binary nb
    for j in (nb_v_bits-a)*['0']+list(bin(val)[2:]): #loops through the reverse ordered list composed of the binary encoding of the value
        if int(j):
            qc.x(nb_i_bits+ pos*5)
            
        pos += 1
        
#This function creates an array of random values that are bellow the threshold indicated, and places nb_sol
#values of (thresh+1) randomely inside the array to act as solutions. This blinds the user to the database
#and they can use it afterwards as a double check on the correct functionong of the circuit. This also has
#the added benefit of making as big an array as the circuit can take.
def array_creation(nb_qb_arr, thresh, nb_sol, extra_sol = []):
    if nb_sol + len(extra_sol) >= 2**(nb_qb_arr-1):
        raise NameError('Number of solution larger than N/2')
        
    data_arr = [randrange(thresh + 1) for x in range(2**nb_qb_arr)]

    i = 0
    while True:
        ind = randrange(len(data_arr))
        if data_arr[ind] < thresh and i < nb_sol:
            data_arr[ind] = thresh+1
            i+=1
        elif data_arr[ind] < thresh and len(extra_sol) != 0:
            data_arr[ind] = extra_sol.pop()
            
        if i >= nb_sol and len(extra_sol) == 0:
            break
            
    return data_arr

In [152]:
nb_qb_arr = 4
nb_sol = 2

thresh = 7
data_arr = array_creation(nb_qb_arr, thresh, nb_sol, extra_sol = [9])


#These numbers are calculated directly from the data array
n = ceil(log(len(data_arr), 2)) #number of index qbits
data_arr += [0]*(2**n - len(data_arr)) #creates the rest of the array so that there are no index problems later

len_b =  ceil(log(max(data_arr)+1, 2))#length of the bitstrings to compare
m = 5*len_b #number of qbits needed for the comparision

i_reg = QuantumRegister(n, 'd') #Creates the data registry
val_reg = QuantumRegister(m-1, 't') #Creates the template registry
anc = QuantumRegister(1, 'ancilla') #Creates the ancillary registry for the oracle
cr = ClassicalRegister(n, 'c') #Creates the classical bit measurment output

qc = QuantumCircuit(i_reg, val_reg, anc, cr) #Makes the circuit with these qbits as input

encode = encode_qc(n,len_b,data_arr).to_gate()
encode.label = "Encoding"

####### State initialization #########
qc.h(range(n))

#initiation of ancillary to |->
qc.x(anc)
qc.h(anc)

##### SETTING THE THRESHOLD ######

set_thresh(qc, n, len_b, thresh)

qc.barrier()

#ENCODING
#The arrays are made with list comprehentions to make something that looks like [0,1,2,n+16,n+11,n+6,n+1], 
#which connects the database (the first values) to the comparators (the values with n+...)
qc.append(encode, [x for x in range(n)]+[n+x*5+1 for x in reversed(range(len_b))])#had to reverse the order how the qbits were entered, as the two parts (database and conparator) work in reverse

comp_qc = n_comparator(len_b)
qc.compose(comp_qc, list(range(n,n+m)),inplace=True) #Compose only adds the circuits together, does not create a gate out of them like append

enc_d = encode.reverse_ops()
enc_d.label = "Encoding_R"
qc.append(enc_d, [x for x in range(n)]+[n+x*5+1 for x in reversed(range(len_b))])

qc.barrier()

#DIFFUSION OP
qc.h(range(n))
qc.x(range(n-1))

#It is difficult to do an actual multi-control-Z gate, so instead we make a MCX surrounded by H gates to transform it into a MCZ
qc.x(n-1)
qc.h(n-1)
qc.mct(list(range(n-1)), n-1) 
qc.h(n-1)
qc.x(n-1)

#The other symmetric side of the diffusor
qc.x(range(n-1))
qc.h(range(n))
qc.barrier()

qc.measure([x for x in range(n)],cr)

qc.draw(fold=-1)

┌───┐      ░ ┌───────────┐                           ┌─────────────┐ ░ ┌───┐┌───┐          ┌───┐┌───┐      ░ ┌─┐         
      d_0: ┤ H ├──────░─┤0          ├───────────────────────────┤0            ├─░─┤ H ├┤ X ├───────■──┤ X ├┤ H ├──────░─┤M├─────────
           ├───┤      ░ │           │                           │             │ ░ ├───┤├───┤       │  ├───┤├───┤      ░ └╥┘┌─┐      
      d_1: ┤ H ├──────░─┤1          ├───────────────────────────┤1            ├─░─┤ H ├┤ X ├───────■──┤ X ├┤ H ├──────░──╫─┤M├──────
           ├───┤      ░ │           │                           │             │ ░ ├───┤├───┤       │  ├───┤├───┤      ░  ║ └╥┘┌─┐   
      d_2: ┤ H ├──────░─┤2          ├───────────────────────────┤2            ├─░─┤ H ├┤ X ├───────■──┤ X ├┤ H ├──────░──╫──╫─┤M├───
           ├───┤      ░ │           │                           │             │ ░ ├───┤├───┤┌───┐┌─┴─┐├───┤├───┤┌───┐ ░  ║  ║ └╥┘┌─┐
      d_3: ┤ H ├──────░─┤3          ├───────────────────────────┤3            ├─░─┤ H ├┤ X ├┤ H ├┤ X ├┤ H ├┤ X ├┤ H ├─░──╫──╫──╫─┤M├
           └───┘      ░ │           │┌────────┐     ┌──────────┐│             │ ░ └───┘└───┘└───┘└───┘└───┘└───┘└───┘ ░  ║  ║  ║ └╥┘
      t_0: ───────────░─┤           ├┤0       ├─────┤0         ├┤             ├─░─────────────────────────────────────░──╫──╫──╫──╫─
                      ░ │           ││        │     │          ││             │ ░                                     ░  ║  ║  ║  ║ 
      t_1: ───────────░─┤7          ├┤1       ├─────┤1         ├┤7            ├─░─────────────────────────────────────░──╫──╫──╫──╫─
                      ░ │           ││        │     │          ││             │ ░                                     ░  ║  ║  ║  ║ 
      t_2: ───────────░─┤           ├┤2       ├─────┤2         ├┤             ├─░─────────────────────────────────────░──╫──╫──╫──╫─
                      ░ │           ││        │     │          ││             │ ░                                     ░  ║  ║  ║  ║ 
      t_3: ───────────░─┤           ├┤3       ├──■──┤3         ├┤             ├─░─────────────────────────────────────░──╫──╫──╫──╫─
                      ░ │           ││        │  │  │          ││             │ ░                                     ░  ║  ║  ║  ║ 
      t_4: ───────────░─┤           ├┤4       ├──┼──┤4         ├┤             ├─░─────────────────────────────────────░──╫──╫──╫──╫─
           ┌───┐      ░ │           ││        │  │  │          ││             │ ░                                     ░  ║  ║  ║  ║ 
      t_5: ┤ X ├──────░─┤           ├┤5       ├──┼──┤5         ├┤             ├─░─────────────────────────────────────░──╫──╫──╫──╫─
           └───┘      ░ │           ││        │  │  │          ││             │ ░                                     ░  ║  ║  ║  ║ 
      t_6: ───────────░─┤6 Encoding ├┤6       ├──┼──┤6         ├┤6 Encoding_R ├─░─────────────────────────────────────░──╫──╫──╫──╫─
                      ░ │           ││        │  │  │          ││             │ ░                                     ░  ║  ║  ║  ║ 
      t_7: ───────────░─┤           ├┤7       ├──┼──┤7         ├┤             ├─░─────────────────────────────────────░──╫──╫──╫──╫─
                      ░ │           ││        │  │  │          ││             │ ░                                     ░  ║  ║  ║  ║ 
      t_8: ───────────░─┤           ├┤8       ├──┼──┤8         ├┤             ├─░─────────────────────────────────────░──╫──╫──╫──╫─
                      ░ │           ││        │  │  │          ││             │ ░                                     ░  ║  ║  ║  ║ 
      t_9: ───────────░─┤           ├┤9  Comp ├──┼──┤9  Comp_R ├┤             ├─░─────────────────────────────────────░──╫──╫──╫──╫─
           ┌───┐      ░ │           ││        │  │  │          ││             │ ░                                     ░  ║  ║  ║  ║ 
     t_10: ┤ X ├──────░─┤           ├┤10      ├──┼──┤10        ├┤             ├─░─────────────────────────────────────░──╫──╫──╫──╫─
           └───┘     

In [153]:
############# CIRCUIT SIMULATION #################
aer_sim = Aer.get_backend('aer_simulator')
aer_sim.set_options(device='CPU')
transpiled_qc = transpile(qc, aer_sim)

shots = 2**(n+7) #we repeat the simulation 2048 times
job = aer_sim.run(transpiled_qc, shots=shots)

############# CODE FOR PLOTTING ####################
hist = job.result().get_counts() #simulation output
sort_hist = sorted(hist.items()) #So that the plotting puts everything in the same increasing order of basis
n_hist = {k:v for k,v in sort_hist}

results = {'val':n_hist.keys(),'count':n_hist.values()} #change the formatting of the data to match plotly

fig = px.bar(results, x="val", y="count", text="count")
fig.update_traces(texttemplate='%{text:.2s}', textposition='outside')
fig.update_layout(uniformtext_minsize=6, uniformtext_mode='show')
fig.show()

############## CODE FOR LISTING THE RESULTS #####################
trig = []
max_c = max(results['count'])
for i,j in zip(results['val'],results['count']):
    if int(j) > max_c-0.33*max_c: #only consideres values that are higher than one third less
        trig.append({'index_dec':int(i,2),'index_bin':i,'counts':j})

for t in trig:
    print(f"Index {t['index_dec']} holds the value {data_arr[t['index_dec']]}, which is above the chosen threshold.")

Index 4 holds the value 9, which is above the chosen threshold.
Index 7 holds the value 8, which is above the chosen threshold.
Index 13 holds the value 8, which is above the chosen threshold.


# Experimenting with Backends

### Links and ressources that I found

* [This link](https://nbviewer.org/github/Qiskit/qiskit-tutorials/tree/master/tutorials/noise/) brings you to the qiskit tutorial jupyter notebook storage. Note worthy for here is the series of tutorials that touch on [simulators](https://nbviewer.org/github/Qiskit/qiskit-tutorials/tree/master/tutorials/simulators/), specifically [this one](https://nbviewer.org/github/Qiskit/qiskit-tutorials/blob/master/tutorials/simulators/1_aer_provider.ipynb) for the basics and [this one](https://nbviewer.org/github/Qiskit/qiskit-tutorials/blob/master/tutorials/simulators/2_device_noise_simulation.ipynb) for specific implementation. This [medium article](https://medium.com/qiskit/qiskit-aer-d09d0fac7759) explains very basically how these work.

* [This article](https://medium.com/qiskit/qiskit-backends-what-they-are-and-how-to-work-with-them-fb66b3bd0463) is a explanation on qiskit backends. In relation to the things discussed above.

* [This](https://qiskit.org/documentation/stubs/qiskit.dagcircuit.DAGCircuit.html) is the documentation for the Direct Acyclic Graphs Circuits (DAGCircuits). These allow the calculation of the number of operations made on the transpiled circuit (so the simulations)

* [This](https://quantumcomputing.stackexchange.com/questions/18505/what-are-the-runtimes-of-the-different-modes-of-qiskits-mcx) is a discussion about circuit depth and the use of 'noancilla'.

In [9]:
from qiskit.dagcircuit import DAGCircuit
from qiskit.converters import circuit_to_dag
from qiskit.compiler import transpile

In [154]:
for backend in Aer.backends():
    print(backend.name())

aer_simulator
aer_simulator_statevector
aer_simulator_statevector_gpu
aer_simulator_density_matrix
aer_simulator_density_matrix_gpu
aer_simulator_stabilizer
aer_simulator_matrix_product_state
aer_simulator_extended_stabilizer
aer_simulator_unitary
aer_simulator_unitary_gpu
aer_simulator_superop
qasm_simulator
statevector_simulator
unitary_simulator
pulse_simulator


## Testing Area

nb of mcx is linked to values in database + length of database, e.g.: 

17 (10001) requires less mcx than 14 (01110) because there is one less bit to encode

So best way to benchmark this is to have the max length of array allowed and highest numbers only, so that we need to encode as many values as possible.

In [155]:
def circ_characteristics(data_arr, thresh):
    #These numbers are calculated directly from the data array
    n = ceil(log(len(data_arr), 2)) #number of index qbits
    len_b =  ceil(log(max(data_arr)+1, 2))#length of the bitstrings to compare
    m = 5*len_b #number of qbits needed for the comparision

    i_reg = QuantumRegister(n, 'd') #Creates the data registry
    val_reg = QuantumRegister(m-1, 't') #Creates the template registry
    anc = QuantumRegister(1, 'ancilla') #Creates the ancillary registry for the oracle
    cr = ClassicalRegister(n, 'c') #Creates the classical bit measurment output

    qc = QuantumCircuit(i_reg, val_reg, anc, cr) #Makes the circuit with these qbits as input

    encode = encode_qc(n,len_b,data_arr).to_gate()
    encode.label = "Encoding database"

    ####### State initialization #########
    qc.h(range(n))

    #initiation of ancillary to |->
    qc.x(anc)
    qc.h(anc)

    ##### SETTING THE THRESHOLD ######

    set_thresh(qc, n, len_b, thresh)

    qc.barrier()

    #ENCODING
    #The arrays are made with list comprehentions to make something that looks like [0,1,2,n+16,n+11,n+6,n+1], 
    #which connects the database (the first values) to the comparators (the values with n+...)
    qc.append(encode, [x for x in range(n)]+[n+x*5+1 for x in reversed(range(len_b))])#had to reverse the order how the qbits were entered, as the two parts (database and conparator) work in reverse

    comp_qc = n_comparator(len_b).to_gate()
    comp_qc.label = "Comparator"

    qc.append(comp_qc, range(n,n+m))

    qc.append(encode.reverse_ops(), [x for x in range(n)]+[n+x*5+1 for x in reversed(range(len_b))])

    qc.barrier()

    #DIFFUSION OP
    qc.h(range(n))
    qc.x(range(n-1))

    #It is difficult to do an actual multi-control-Z gate, so instead we make a MCX surrounded by H gates to transform it into a MCZ
    qc.x(n-1)
    qc.h(n-1)
    qc.mct(list(range(n-1)), n-1) 
    qc.h(n-1)
    qc.x(n-1)

    #The other symmetric side of the diffusor
    qc.x(range(n-1))
    qc.h(range(n))
    qc.barrier()

    qc.measure([x for x in range(n)],cr)
    
    ######### SIMULATION CIRCUIT CREATION ########
    aer_sim = Aer.get_backend('aer_simulator')
    transpiled_qc = transpile(qc, aer_sim)
    #we don't need to run the simulation for benchmark
    
    dag_circuit = circuit_to_dag(transpiled_qc)

    
    return transpiled_qc, dag_circuit.count_ops(), dag_circuit.num_qubits()

# This function prints some information about the simulated circuit
def print_info_simCirc(dict_out, qbit_out):
    print("Operation \tNumber\n")
    for i,j in dict_out.items():
        if len(i) < 4:
            print(f"{i} \t\t{j}")
        else:
            print(f"{i} \t{j}")

    print(f"\n\nNumber of Qbits used: {qbit_out}\n\n")

Let's use the benchmark function to see a few different values. Here we choose the maximum for four bits ($2^4=16$)  for the array and the max value.

In [156]:
data_arr = [15]*16
thresh = 15

tr_qc, dict_out, qbit_out = circ_characteristics(data_arr, thresh)

print_info_simCirc(dict_out, qbit_out)

Operation 	Number

h 		4
x 		112
u2 		7
barrier 	3
mcx 		129
ccx 		34
cx 		1
u1 		2
measure 	4


Number of Qbits used: 24




And print the simulated circuit with the above noted operations

In [157]:
tr_qc.draw(fold=-1) #can draw the simulated circuit

global phase: 0
              ┌───┐     ░ ┌───┐                    ┌───┐                    ┌───┐                    ┌───┐                    ┌───┐                    ┌───┐                    ┌───┐                    ┌───┐                    ┌───┐                    ┌───┐                    ┌───┐                    ┌───┐                    ┌───┐                    ┌───┐                    ┌───┐                    ┌───┐                                                                                                                                            ┌───┐                    ┌───┐                    ┌───┐                    ┌───┐                    ┌───┐                    ┌───┐                    ┌───┐                    ┌───┐                    ┌───┐                    ┌───┐                    ┌───┐                    ┌───┐                    ┌───┐                    ┌───┐                    ┌───┐                    ┌───┐ ░ ┌─────────┐     ┌──────────┐ ░ ┌─┐         
      d_0: ───┤ H ├─────░─┤ X ├──■────■────■────■──┤ X ├──■────■────■────■──┤ X ├──■────■────■────■──┤ X ├──■────■────■────■──┤ X ├──■────■────■────■──┤ X ├──■────■────■────■──┤ X ├──■────■────■────■──┤ X ├──■────■────■────■──┤ X ├──■────■────■────■──┤ X ├──■────■────■────■──┤ X ├──■────■────■────■──┤ X ├──■────■────■────■──┤ X ├──■────■────■────■──┤ X ├──■────■────■────■──┤ X ├──■────■────■────■──┤ X ├──■────■────■────■────────────────────────────────────────────────────────────────────────────────────────────────────────■────■────■────■──┤ X ├──■────■────■────■──┤ X ├──■────■────■────■──┤ X ├──■────■────■────■──┤ X ├──■────■────■────■──┤ X ├──■────■────■────■──┤ X ├──■────■────■────■──┤ X ├──■────■────■────■──┤ X ├──■────■────■────■──┤ X ├──■────■────■────■──┤ X ├──■────■────■────■──┤ X ├──■────■────■────■──┤ X ├──■────■────■────■──┤ X ├──■────■────■────■──┤ X ├──■────■────■────■──┤ X ├──■────■────■────■──┤ X ├─░─┤ U2(0,0) ├──■──┤ U2(π,-π) ├─░─┤M├─────────
              ├───┤     ░ ├───┤  │    │    │    │  └───┘  │    │    │    │  ├───┤  │    │    │    │  └───┘  │    │    │    │  ├───┤  │    │    │    │  └───┘  │    │    │    │  ├───┤  │    │    │    │  └───┘  │    │    │    │  ├───┤  │    │    │    │  └───┘  │    │    │    │  ├───┤  │    │    │    │  └───┘  │    │    │    │  ├───┤  │    │    │    │  └───┘  │    │    │    │  ├───┤  │    │    │    │  └───┘  │    │    │    │                                                                                                        │    │    │    │  └───┘  │    │    │    │  ├───┤  │    │    │    │  └───┘  │    │    │    │  ├───┤  │    │    │    │  └───┘  │    │    │    │  ├───┤  │    │    │    │  └───┘  │    │    │    │  ├───┤  │    │    │    │  └───┘  │    │    │    │  ├───┤  │    │    │    │  └───┘  │    │    │    │  ├───┤  │    │    │    │  └───┘  │    │    │    │  ├───┤  │    │    │    │  └───┘  │    │    │    │  ├───┤ ░ ├─────────┤  │  ├──────────┤ ░ └╥┘┌─┐      
      d_1: ───┤ H ├─────░─┤ X ├──■────■────■────■─────────■────■────■────■──┤ X ├──■────■────■────■─────────■────■────■────■──┤ X ├──■────■────■────■─────────■────■────■────■──┤ X ├──■────■────■────■─────────■────■────■────■──┤ X ├──■────■────■────■─────────■────■────■────■──┤ X ├──■────■────■────■─────────■────■────■────■──┤ X ├──■────■────■────■─────────■────■────■────■──┤ X ├──■────■────■────■─────────■────■────■────■────────────────────────────────────────────────────────────────────────────────────────────────────────■────■────■────■─────────■────■────■────■──┤ X ├──■────■────■────■─────────■────■────■────■──┤ X ├──■────■────■────■─────────■────■────■────■──┤ X ├──■────■────■────■─────────■────■────■────■──┤ X ├──■────■────■────■─────────■────■────■────■──┤ X ├──■────■────■────■─────────■────■────■────■──┤ X ├──■────■────■────■─────────■────■────■────■──┤ X ├──■────■────■────■─────────■────■────■────■──┤ X ├─░─┤ U2(0,0) ├──■──┤ U2(π,-π) ├─░──╫─┤M├──────
              ├───┤     ░ ├───┤  │    │    │    │         │    │    │    │  └───┘  │    │   

## Characterising  the circuit

### Creating the dataset
Here we create the benchmark data by looping through all possibilities of value and index range. In this case we loop from $2^2$ to $2^7$ as a value range.

In [171]:
range_qbit = [3, 11]
x = [2**x for x in range(range_qbit[0],range_qbit[1])]
y = [2**x for x in range(range_qbit[0],range_qbit[1])]

plot={
    "q":[],
    "mcx":[],
    "cx":[],
    "x":[],
     }

for i in x:
    
    nb_q = []
    nb_mcx = []
    nb_cx = []
    nb_x = []
    
    for j in y:
        
        #make the simulation circuit
        tr_qc, dict_out, qbit_out = circ_characteristics([j-1]*i, j-1)
        
        #number of qbits
        nb_q.append(qbit_out)
        
        #number of mcx
        if 'mcx' in dict_out:
            nb_mcx.append(dict_out['mcx'])
        else:
            nb_mcx.append(dict_out['mcx_gray'])
            
        #number of cx
        nb_cx.append(dict_out['cx'])
        
        #number of x
        nb_x.append(dict_out['x'])
        
    plot["q"].append(nb_q)
    plot["mcx"].append(nb_mcx)
    plot["cx"].append(nb_cx)
    plot["x"].append(nb_x)


### Plot Using go.Surface()

In [180]:
fig = make_subplots(
    rows=2, cols=2,
    specs=[[{'type': 'surface'}, {'type': 'surface'}],
           [{'type': 'surface'}, {'type': 'surface'}]],
    subplot_titles=[
        "Number of Qbits",
        "Number of MCX",
        "Number of CX",
        "Number of X",
    ]
)

s1 = fig.add_trace(
    go.Surface(z=plot['q'], x=np.log(x), y=np.log(y),
              colorbar={"lenmode":'fraction', 'len':0.5, 'thickness':20, 'x':-0.2, 'y':0.8}),
    row=1, col=1)

s2 = fig.add_trace(
    go.Surface(z=plot['mcx'], x=np.log(x), y=np.log(y),
              colorbar={"lenmode":'fraction', 'len':0.5, 'thickness':20, 'x':1.1, 'y':0.8}),
    row=1, col=2)

s3 = fig.add_trace(
    go.Surface(z=plot['cx'], x=np.log(x), y=np.log(y),
              colorbar={"lenmode":'fraction', 'len':0.5, 'thickness':20, 'x':-0.2, 'y':0.22}),
    row=2, col=1)

s4 = fig.add_trace(
    go.Surface(z=plot['x'], x=np.log(x), y=np.log(y),
              colorbar={"lenmode":'fraction', 'len':0.5, 'thickness':20, 'x':1.1, 'y':0.22}),
    row=2, col=2)

name = "Benchmark of the simulation: impact of database and value size on operation numbers (x: len array; y: max val)"

# default is (x=1.25,y=1.25,z=1.25)
camera = {"eye":{"x":-1.6, "y":-1.6, "z":0.6}}

fig.update_layout(#title=name,
                  height=1000,
                  width=1000
                 )

fw = go.FigureWidget(fig)
# print(fw.layout) #prints information about the layout of the plot

with fw.batch_update():
    fw.layout.scene1.camera.eye=camera["eye"]
    fw.layout.scene2.camera.eye=camera["eye"]
    fw.layout.scene3.camera.eye=camera["eye"]
    fw.layout.scene4.camera.eye=camera["eye"]
    
fw.show()

if not os.path.exists("plots/complex_graphs"):
        os.makedirs("plots/complex_graphs")
        
fw.write_image("plots/complex_graphs/Surf2x2.pdf",width=1000, height=1000,)
fw.write_html("plots/complex_graphs/Surf2x2.html")
        

### Plot Using go.Heatmap()

#### Discrete

In [182]:
fig = make_subplots(
    rows=2, cols=2,
    specs=[[{'type': 'heatmap'}, {'type': 'heatmap'}],
           [{'type': 'heatmap'}, {'type': 'heatmap'}]],
    subplot_titles=[
        "Number of Qbits",
        "Number of MCX",
        "Number of CX",
        "Number of X",
    ]
)

fig.add_trace(
    go.Heatmap(z=plot["q"],
                x=np.log2(x),
                y=np.log2(y),
                colorbar={"title":'#Qbits', "lenmode":'fraction', 'len':0.4, 'thickness':20, 'x':-0.2, 'y':0.8},
               ),
    row=1, col=1)

fig.add_trace(
    go.Heatmap(z=plot["mcx"],
                colorbar={"title":'#MCX', "lenmode":'fraction', 'len':0.4, 'thickness':20, 'x':1.1, 'y':0.8} ,
                x=np.log2(x),
                y=np.log2(y),
               ),
    row=1, col=2)

fig.add_trace(
    go.Heatmap(z=plot["cx"],
                colorbar={"title":'#CX', "lenmode":'fraction', 'len':0.4, 'thickness':20, 'x':-0.2, 'y':0.22} ,
                x=np.log2(x),
                y=np.log2(y),
               ),
    row=2, col=1)

fig.add_trace(
    go.Heatmap(z=plot["x"],
                colorbar={"title":'#X', "lenmode":'fraction', 'len':0.4, 'thickness':20, 'x':1.1, 'y':0.22} ,
                x=np.log2(x),
                y=np.log2(y),
               ),
    row=2, col=2)

name = "Benchmark of the simulation: impact of database and value size on operation numbers (x: len array; y: max val)"

fig.update_xaxes(title_text="Length of array")
fig.update_yaxes(title_text="Max value of array")

fig.update_layout(title=name,
                  height=800,
                  width=1000
                 )


In [185]:
fig = go.Figure(data=go.Heatmap(z=(np.array(plot["mcx"])+np.array(plot["x"])),
                colorbar={"title":'# of mcx'} ,
                x=np.log2(x),
                y=np.log2(y),
               ))

fig.update_xaxes(tickformat=',d')
fig.update_yaxes(tickformat=',d')

fig.update_layout(
#     title="Benchmark of the simulation: impact of database and value size on MCX + CX operation number",
    xaxis_title="Max value of array",
    yaxis_title="Length of array (splicing)",
    title_x=0.5
)

fig.show()

if not os.path.exists("plots/complex_graphs"):
        os.makedirs("plots/complex_graphs")
        
fig.write_image("plots/complex_graphs/mcx_x_heatmap.pdf",width=1000, height=600,)
fig.write_html("plots/complex_graphs/mcx_x_heatmap.html")
        

#### Interpolated (smoothed)

In [163]:
fig = make_subplots(
    rows=2, cols=2,
    specs=[[{'type': 'heatmap'}, {'type': 'heatmap'}],
           [{'type': 'heatmap'}, {'type': 'heatmap'}]],
    subplot_titles=[
        "Number of Qbits",
        "Number of MCX",
        "Number of CX",
        "Number of X",
    ]
)

fig.add_trace(
    go.Heatmap(z=plot["q"],
                x=np.log2(x),
                y=np.log2(y),
                colorbar={"title":'#Qbits', "lenmode":'fraction', 'len':0.4, 'thickness':20, 'x':-0.2, 'y':0.8},
                zsmooth= 'best'
               ),
    row=1, col=1)

fig.add_trace(
    go.Heatmap(z=plot["mcx"],
                colorbar={"title":'#MCX', "lenmode":'fraction', 'len':0.4, 'thickness':20, 'x':1.1, 'y':0.8} ,
                x=np.log2(x),
                y=np.log2(y),
                zsmooth= 'best'
               ),
    row=1, col=2)

fig.add_trace(
    go.Heatmap(z=plot["cx"],
                colorbar={"title":'#CX', "lenmode":'fraction', 'len':0.4, 'thickness':20, 'x':-0.2, 'y':0.22} ,
                x=np.log2(x),
                y=np.log2(y),
                zsmooth= 'best'
               ),
    row=2, col=1)

fig.add_trace(
    go.Heatmap(z=plot["x"],
                colorbar={"title":'#X', "lenmode":'fraction', 'len':0.4, 'thickness':20, 'x':1.1, 'y':0.22} ,
                x=np.log2(x),
                y=np.log2(y),
                zsmooth= 'best'
               ),
    row=2, col=2)

name = "Benchmark of the simulation: impact of database and value size on operation numbers (x: len array; y: max val)"

fig.update_xaxes(title_text="Length of array")
fig.update_yaxes(title_text="Max value of array")

fig.update_layout(title=name,
                  height=800,
                  width=1000
                 )


Unused code with different plotting choices. To check them out, please transform the cells back into "Code" type.

### Plot Using px.imshow() for heatmap

In [ ]:
fig = px.imshow(plot['mcx'],
                x=np.log2(x),
                y=np.log2(y),
                labels={"color":"# of mcx",
                       "x":"Length of array, logarithmic plot (2**x))",
                       "y":"Max value of array, logarithmic plot (2**x))",
                       }
               )

fig.update_xaxes(tickformat=',d')
fig.update_yaxes(tickformat=',d',autorange=True)

fig.update_layout(
    title="Benchmark of the simulation: impact of database and value size on MCX operation number",
    width=800, height=700
)

fig.show()